pip/pip-415.md

```diff
@@ -0,0 +1,112 @@
+# PIP-415: Support getting message ID by offset
+
+# Background knowledge
+
+
+- In [PIP-70](https://github.com/apache/pulsar/wiki/PIP-70%3A-Introduce-lightweight-broker-entry-metadata), we introduced `Broker Entry Metadata` containing an `index` field - a continuous sequence identifier for messages within a topic partition.
+This sequence ID simplifies message sequence management compared to the composite MessageId (ledgerId+entryId+batchIndex) which becomes discontinuous across ledgers.
+- In [PIP-90](https://github.com/apache/pulsar/wiki/PIP-90%3A-Expose-broker-entry-metadata-to-the-client), we exposed this metadata via the `Message.getIndex()` API. 
+- In [PR-6331](https://github.com/apache/pulsar/pull/6331), we added a new `get-message-by-id` cmd into pulsar-admin.
+
+After these three works, we can now obtain the offset of a message by its message id:
+
+1. Get the message by id using `get-message-by-id` cmd
+2. Get the index of the message using `Message.getIndex()`
+
+But we cannot obtain the message id by offset. Then we need to add a new API to get the message id by offset.
+
+# Motivation
+
```

Comment:
I think we should not blindly compare pulsar and Rocketmq. This is not convincing at all. I think we can look at it from our own perspective.


Motivation is written relatively simply. I feel that some index planning can be added. Currently, the index function is open ( brokerEntryMetadataInterceptors=org.apache.pulsar.common.intercept.AppendIndexMetadataInterceptor
), but some functions are actually imperfect, such as:

Get MessageId through index
Get the corresponding index through MessageId
This time, the function of getting MessageId through index will be improved first